File: app/src/main/java/vn/loitp/app/activity/pattern/observerpattern/ObserverPatternActivity.java
package vn.loitp.app.activity.pattern.observerpattern;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import loitp.basemaster.R;
import vn.loitp.core.base.BaseActivity;
import vn.loitp.views.LToast;

//https://viblo.asia/p/android-design-patterns-the-observer-pattern-WAyK8xqpKxX
public class ObserverPatternActivity extends BaseActivity implements View.OnClickListener, RepositoryObserver {
    private Subject mUserDataRepository;
    private TextView tv;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        tv = (TextView) findViewById(R.id.tv);

        mUserDataRepository = UserDataRepository.getInstance();
        mUserDataRepository.registerObserver(this);

        findViewById(R.id.bt).setOnClickListener(this);
    }

    @Override
    protected boolean setFullScreen() {
        return false;
    }

    @Override
    protected String setTag() {
        return getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_observer_pattern;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.bt:
                tv.setText("Loading...");
                UserDataRepository.getInstance().getNewDataFromRemote();
                break;
        }
    }

    @Override
    public void onUserDataChanged(String fullname, int age) {
        LToast.show(activity, "onUserDataChanged");
        tv.setText(fullname + " - " + age);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mUserDataRepository.removeObserver(this);
    }
}

File: app/src/main/java/vn/loitp/app/activity/pattern/observerPattern/ObserverPatternActivity.kt
package vn.loitp.app.activity.pattern.observerPattern

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import androidx.core.view.isVisible
import com.loitpcore.annotation.IsFullScreen
import com.loitpcore.annotation.LogTag
import com.loitpcore.core.base.BaseFontActivity
import com.loitpcore.core.utilities.LUIUtil
import kotlinx.android.synthetic.main.activity_pattern_observer.*
import vn.loitp.app.R

// https://viblo.asia/p/android-design-patterns-the-observer-pattern-WAyK8xqpKxX

@LogTag("ObserverPatternActivity")
@IsFullScreen(false)
class ObserverPatternActivity : BaseFontActivity(), View.OnClickListener, RepositoryObserver {
    private var mUserDataRepository: Subject? = null

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_pattern_observer
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mUserDataRepository = UserDataRepository.getInstance()
        mUserDataRepository?.registerObserver(this)

        setupViews()
    }

    private fun setupViews() {
        lActionBar.apply {
            LUIUtil.setSafeOnClickListenerElastic(
                view = this.ivIconLeft,
                runnable = {
                    onBackPressed()
                }
            )
            this.ivIconRight?.setImageResource(R.color.transparent)
            this.viewShadow?.isVisible = true
            this.tvTitle?.text = ObserverPatternActivity::class.java.simpleName
        }
        bt.setOnClickListener(this)
    }

    @SuppressLint("SetTextI18n")
    override fun onClick(v: View) {
        when (v) {
            bt -> {
                textView.text = "Loading..."
                UserDataRepository.getInstance()?.getNewDataFromRemote()
            }
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onUserDataChanged(fullName: String, age: Int) {
        showShortInformation("onUserDataChanged $fullName - $age")
        textView.text = "$fullName - $age"
    }

    override fun onDestroy() {
        super.onDestroy()
        mUserDataRepository?.removeObserver(this)
    }
}
